### PR TITLE
Asyncio tarantool queue

### PR DESCRIPTION
**Pull request by jetuuuu**
Implemented queue interface. This [issue](https://github.com/tarantool/go-tarantool/issues/31).

### CODE REVIEW

File: queue.go [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+		ret["pri"] = opts.Pri
+	}
+
+	return ret
+}
+
+func newQueue(conn *Connection, name string, cfg queueCfg) (Queue, error) {
+	var q queue
+	cmd := fmt.Sprintf("queue.create_tube('%s', '%s', %s)", name, cfg.Type(), cfg.String())
+	fut := conn.EvalAsync(cmd, []interface{}{})
```
Comment (funny-falcon) [inline queue.go]: Why EvalAsync if you wait in the next row?

File: queue.go [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+	}
+
+	return task, err
+}
+
+func (q queue) Take() (*Task, error) {
+	return q.take(nil)
+}
+
+func (q queue) TakeWithTimeout(timeout time.Duration) (*Task, error) {
```
Comment (funny-falcon) [inline queue.go]: What will happen, if timeout is greater than request timeout on connection.

**Comment by jetuuuu** [comment]
Ok, I understood the main problem. I'll think about the implementation and correct it.

**Comment by funny-falcon** [comment]
Probably, at this level of request timeout implementation, it is more documentation issue than implementation.
I do not demand you to implement real per request timeout. (though, I wouldn't mind if you do).

**Comment by funny-falcon** [comment]
Concerning current implementation, you may check in Take and TakeWithTimeout that Connection uses compatible request timeout, and return error otherwise.

**Comment by jetuuuu** [comment]
Hi, look at my last commit when you have a time.

**Comment by funny-falcon** [comment]
Much better.
But still no documentation in code comments.
Look at https://godoc.org/github.com/tarantool/go-tarantool
What whould you like to read there about Queue if it were written by someone else ?
When your fellow will ask to describe Queue, what would you like he could read on this documentation page?

https://blog.golang.org/godoc-documenting-go-code

PS. fill free to open pull requests for improving current documentation.

**Comment by jetuuuu** [comment]
Hi, sorry, I was busy at work, could not make corrections earlier.

**Comment by funny-falcon** [comment]
Excuse me for wasting your time, but I have a bit more suggestions (after talk with a friend who uses go-tarantool and python connector):
1. please, move queue into subfolder, so it will have separate `import` part. As a benefit, `tarantool.QueueOpts` and `tarantool.QueueCfq` will be shortened to `queue.Opts` and `queue.Cfg`. (`tarantool.FIFO_QUEUE` => `queue.FIFO`, etc...)
2. I don't see benefits from separating QueueCfg and TtlQueueCfg. It should be one `queue.Cfg` with dispatching by cfg.Kind.
3. `TakeWithTimeout` should be `TakeTimeout` (see `net.DialTimeout`).
`PutWithConfig` => `PutWithOpts` (or `q.Put(data)` => `q.Put(data, queue.Opts{})`, and I don't know, which is better)

What do you think about this comments? Would you mind to apply them?

**Comment by jetuuuu** [comment]
> I don't see benefits from separating QueueCfg and TtlQueueCfg. It should be one queue.Cfg with dispatching by cfg.Kind.

Do you mean one type with all possible fields? For example:

```
type QueueCfg struct {
        Temporary   bool // if true, the contents do not persist on disk
	IfNotExists bool // if true, no error will be returned if the tube already exists
	Kind        queueType
	Pri   int           // task priorities
	Ttl   time.Duration // task time to live
	Ttr   time.Duration // task time to execute
	Delay time.Duration // delayed execution
}
```

**Comment by shveenkov** [comment]
> Do you mean one type with all possible fields?
yes, or like this:

```golang

type QueueCfg struct {
        Temporary   bool // if true, the contents do not persist on disk
	IfNotExists bool // if true, no error will be returned if the tube already exists
	Kind        queueType
	QueueOpts
}
```

File: queue/queue.go [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+	return q.put(data, cfg.toMap())
+}
+
+func (q queue) put(p ...interface{}) (*Task, error) {
+	var (
+		params []interface{}
+		task   *Task
+	)
+	params = append(params, p...)
+	resp, err := q.conn.c.Call(q.cmd["put"], params)
```
Comment (funny-falcon) [inline queue/queue.go]: I don't demand this change. If you like it then apply. If you don't, I wouldn't mind.

````go
type cnt int                                                                                                                                                                                                                              
func (c *cnt) inc(add bool) {                                                                                                                                                                                                             
    if add {                                                                                                                                                                                                                              
        *c++                                                                                                                                                                                                                              
    }                                                                                                                                                                                                                                     
}                                                                                                                                                                                                                                         
                                                                                                                                                                                                                                          
func (o Opts) EncodeMsgpack(e msgpack.Encoder) error {                                                                                                                                                                                    
    cnt := cnt(0)                                                                                                                                                                                                                         
    cnt.add(opts.Ttl != 0)                                                                                                                                                                                                                
    cnt.add(opts.Ttr != 0)                                                                                                                                                                                                                
    cnt.add(opts.Delay != 0)                                                                                                                                                                                                              
    cnt.add(opts.Pri != 0)                                                                                                                                                                                                                
    if cnt == 0 {                                                                                                                                                                                                                         
        return e.EncodeNil()                                                                                                                                                                                                              
    }                                                                                                                                                                                                                                     
    e.EncodeMapLen(int(cnt))                                                                                                                                                                                                       
    if opts.Ttl != 0 {                                                                                                                                                                                                                    
        e.EncodeString("ttl")                                                                                                                                                                                                             
        e.EncodeFloat64(opts.Ttl.Seconds())                                                                                                                                                                                               
    }                                                                                                                                                                                                                                     
    if opts.Ttr != 0 {                                                                                                                                                                                                                    
        e.EncodeString("ttr")                                                                                                                                                                                                             
        e.EncodeFloat64(opts.Ttr.Seconds())                                                                                                                                                                                               
    }                                                                                                                                                                                                                                     
    if opts.Delay != 0 {                                                                                                                                                                                                                  
        e.EncodeString("delay")                                                                                                                                                                                                           
        e.EncodeFloat64(opts.Delay.Seconds())                                                                                                                                                                                             
    }                                                                                                                                                                                                                                     
    if opts.Pri != 0 {                                                                                                                                                                                                                    
        e.EncodeString("pri")                                                                                                                                                                                                             
        e.EncodeInt(opts.Pri)                                                                                                                                                                                                             
    }                                                                                                                                                                                                                                     
    return nil                                                                                                                                                                                                                            
}                                                                                                                                                                                                                                         
                                                                                                                                                                                                                                          
func (q queue) Put(data interface{}) (*Task, error) {                                                                                                                                                                                     
    return q.PutWithOpts(data, Opts{})                                                                                                                                                                                                    
}                                                                                                                                                                                                                                         
                                                                                                                                                                                                                                          
// Put data with options (ttl/ttr/pri/delay) to queue. Returns task.                                                                                                                                                                      
func (q queue) PutWithOpts(data interface{}, cfg Opts) (*Task, error) {                                                                                                                                                                   
    var task *Task                                                                                                                                                                                                                        
    params := []interface{}{ data, nil }                                                                                                                                                                                                  
    if cfg != Opts{} {                                                                                                                                                                                                                    
        params[1] = cfg                                                                                                                                                                                                                   
    }                                                                                                                                                                                                                                     
    resp, err := q.conn.c.Call(q.cmd["put"], params)
````

File: README.md [hunk 4]
```diff
@@ -499,3 +500,47 @@ func decodeTuple(d *msgpack.Decoder, v reflect.Value) error {
   up. If `MaxReconnects` is zero, the client will try to reconnect endlessly.
 * `User` - user name to log into Tarantool.
 * `Pass` - user password to log into Tarantool.
+
+## Working-with-queue
+```go
+opts := tarantool.Opts{...}
+conn, err := tarantool.Connect("127.0.0.1:3301", opts)
+
+cfg := tarantool.TtlQueueCfg{
```
Comment (funny-falcon) [inline README.md]: You didn't fix documentation with refactor.

File: queue/queue.go [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+	TakeTimeout(timeout time.Duration) (*Task, error)
+	Drop() error
+	Peek(taskId uint64) (*Task, error)
+	Kick(taskId uint64) (uint64, error)
+	Statistic() (interface{}, error)
+}
+
+type Connection struct {
+	c    *tarantool.Connection
+	opts tarantool.Opts
```
Comment (funny-falcon) [inline queue/queue.go]: Would you like to have `tarantool.Connection.ConfiguredTimeout() time.Duration` ?
Then you wouldn't need `queue.Connection`, and then `queue.NewConn(connection)`.

File: README.md (hunk 4 above)
Comment (jetuuuu) [inline README.md]: Yes, my mistake, sorry ...

File: queue/queue.go (hunk 5 above)
Comment (jetuuuu) [inline queue/queue.go]: Yes, I thought about it, but did not dare to implement. Do you mind if I implement this?

File: queue.go [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+func (q queue) _bury(taskId uint64) (string, error) {
+	return q.produce("bury", taskId)
+}
+
+func (q queue) _release(taskId uint64, cfg QueueOpts) (string, error) {
+	return q.produce("release", taskId, cfg.toMap())
+}
+func (q queue) produce(cmd string, p ...interface{}) (string, error) {
+	var params []interface{}
+	params = append(params, p...)
```
Comment (jetuuuu) [inline queue.go]: Because… I'm stupid (( my mistake

File: queue/queue.go (hunk 5 above)
Comment (funny-falcon) [inline queue/queue.go]: No, I wouldn't mind.

File: queue.go (hunk 6 above)
Comment (funny-falcon) [inline queue.go]: I understand this construction, and why it could be used. It is just not the case.

Comment (jetuuuu) [inline queue.go]: But I can use just `p` as parameter without converting to `[]interface{}`

**Comment by funny-falcon** [comment]
Excuse me that it lasts so long.

Minor remark: I don't like cmdMap. It should be struct. But I may to fix it by myself, cause it is not public interface.

Major remark: I don't like use of Eval. It demands 'execute universe' privilege.
If you don't need to create queue, then there is a little issue to check it's existence: you will fail on access to it. If you really want to check existence, you may get statistics. If you need to create, then you may call queue.create_tube as a function.
So, I suggest interface:
```go
  q := queue.New(conn, "name")
  // optionally check existence and create
  if _, err := q.Statistics(); err != nil {
    if err = q.Create(queue.Cfg{Kind: queue.FIFO}); err != nil {
      panic("Шеф, всё пропало");
    }
  }
```
What do you think about?

**Comment by funny-falcon** [comment]
Also, there is always `q.Create(queue.Cfg(IfNotExists: true, Kind: queue.FIFO))`, if you want simplify code a bit. Btw, may be rename `IfNotExists` to `FailOnExists` and pass its negation as 'if_not_exists'? So that, default value will match `if_not_exists=true`.
Your opinion?

**Comment by jetuuuu** [comment]
> I don't like cmdMap

ok, I can change map to struct :)

> If you really want to check existence, you may get statistics

In this case I will receive a lot of unnecessary information

> Btw, may be rename `IfNotExists` to `FailOnExists`

I followed the lua api … I think we do not need to rename the field and mislead people

>  I don't like use of Eval

But the `Eval` will remain to create the tube.

**Comment by funny-falcon** [comment]
> But the Eval will remain to create the tube.

???? why? `queue.create_tube` is a function and could be called as a function. No?
If client user may create tube, then it could be given 'execute queue.create_tube' right or 'execute universe'.
But if client is restricted to only use queue, then it could be given access only for concrete tube methods.
So, no flexibility is loosed if you call `queue.create_tube` as a function instead of Eval.
Am I missing something?

> I followed the lua api … I think we do not need to rename the field and mislead people

I just thought that default behavior "if_not_exists=true" is more correct, but I could be mistaken.

>> If you really want to check existence, you may get statistics

> In this case I will receive a lot of unnecessary information

```go
func (q Queue) Exists() error {
   return q.Create(Cfg{Kind: "dumb_non_existing_type", IfNotExists: true})
}
```
This will return error if queue doesn't exists, and will not do anything if it exists.
Most of time one simply doesn't need in check for existence: either it is ought to be already created (with migration), or application will create it right now.
I could be mistaken.

PS. I see that you pass `(q queue)` as a value to methods. So if you will change cmdMap to struct, then Queue should contain pointer to struct. Better way is to change method receiver to `(q *queue)`. It is also will cheaper, cause in current way queue should be copied to stack every function call.

PPS. If you already boring, I may merge current state to separate branch, and then make my changes by myself. I really appreciate the work you've already done. If you want to go this way, last request is to squash commits into single one.

**Comment by jetuuuu** [comment]
> If you already boring

No, I'm full of energy :)

> `queue.create_tube` is a function and could be called as a function

`Call("queue.create_tube", []interface{name, type, cfg}` -- return error `unsupported Lua type 'function'`, because msgpack can't represent a Lua function object.

> func (q Queue) Exists() error {
   return q.Create(Cfg{Kind: "dumb_non_existing_type", IfNotExists: true})
}
This will return error if queue doesn't exists, and will not do anything if it exists.

Hmm… trying to create a queue… Do I understand correctly?

**Comment by funny-falcon** [comment]
> because msgpack can't represent a Lua function object.

I see. Its pity. Ok, `func (q queue) Create(Cfg)` could use Eval, cause there is no option.
And I understand than there is no way to check existence without Eval currently. I opened https://github.com/tarantool/queue/issues/58 , but until that lets use Eval in a separate method:
```go
q := queue.New(conn, "myqueue")
// optional check for existence below
var ok bool
var err error
if ok, err = q.Exists(); err == nil && !ok {
    err = q.Create(Cfg{Kind: queue.FIFO})
}
if err != nil {
    panic("Шеф, всё пропало!")
}
```
So, if programmer don't want to use Eval, it could skip call to `q.Exists()`.

PS. you don't pass queue name in `Statistic`:
```go
func (q queue) Statistic() (interface{}, error) {
    resp, err := q.conn.Call(q.cmd["statistics"], []interface{}{q.name})
    ...
}
```

**Comment by jetuuuu** [comment]
> PS. you don't pass queue name in Statistic:
func (q queue) Statistic() (interface{}, error) {
    resp, err := q.conn.Call(q.cmd["statistics"], []interface{}{q.name})
    ...
}

why ? Here it is `[]interface{}{q.name}`.

File: queue/queue.go [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+	var id uint64
+	if err == nil {
+		id = resp.Data[0].([]interface{})[0].(uint64)
+	}
+	return id, err
+}
+
+// Return the number of tasks in a queue broken down by task_state, and the number of requests broken down by the type of request.
+func (q queue) Statistic() (interface{}, error) {
+	resp, err := q.conn.Call(q.cmd["statistics"], []interface{}{})
```
Comment (funny-falcon) [inline queue/queue.go]: Pass the name of a queue as a parameter.
Otherwise you will got statistic for all queues.

**Comment by funny-falcon** [comment]
https://golang.org/doc/effective_go.html#Getters
https://golang.org/doc/effective_go.html#commentary

**Comment by funny-falcon** [comment]
I've pushed my refactor to `dev` branch. Last chance to review it. Tomorrow evening I will merge it into `master` branch.

**Comment by funny-falcon** [comment]
I've merged today into `master` branch.

**Pull request merged**